arrows/vxl/convert_image.cxx

```diff
@@ -0,0 +1,499 @@
+  {
+    // Find the index by multiplying the number of points by the percentile
+    // The number is adjusted by -1 to account for the fact that percentiles
+    // are the number which fall below a value. The +0.5 is to account for
+    // truncation by static cast.
+    unsigned ind =
+      static_cast< unsigned >( sampling_points_minus1 * percentiles[ i ] +
+                               0.5 );
+    dst[ i ] = sorted_samples[ ind ];
+  }
+  return dst;
+}
+
+// ----------------------------------------------------------------------------
+template < typename InputType, typename OutputType >
+void
+percentile_scale_image(
+  vil_image_view< InputType > const& src, vil_image_view< OutputType >& dst,
+  double lower, double upper, unsigned sampling_points,
+  bool ignore_extremes = true )
+{
+  std::vector< double > percentiles( 2, 0.0 );
+  percentiles[ 0 ] = lower;
+  percentiles[ 1 ] = upper;
+
+  std::vector< InputType > percentile_values =
+    get_image_percentiles( src, percentiles, sampling_points,
+                           ignore_extremes );
+
+  OutputType max_val = std::numeric_limits< OutputType >::max();
+
+  InputType lower_bound = percentile_values[ 0 ];
+  InputType upper_bound = percentile_values[ 1 ];
+
+  double scale;
+
+  if( percentile_values[ 1 ] - percentile_values[ 0 ] > 0 )
+  {
+    scale = ( static_cast< double >( max_val ) + 0.5 ) /
+            ( percentile_values[ 1 ] - percentile_values[ 0 ] );
+  }
+  else
+  {
+    scale = static_cast< double >( max_val ) /
+            std::numeric_limits< InputType >::max();
+  }
+
+  auto const ni = src.ni();
+  auto const nj = src.nj();
+  auto const np = src.nplanes();
+  dst.set_size( ni, nj, np );
+
+  // Stretch image to upper and lower percentile bounds
+  vil_transform( src, dst,
+                 [ lower_bound, upper_bound, scale ]( InputType pixel ){
+                   if( pixel < lower_bound )
+                   {
+                     return static_cast< OutputType >( 0 );
+                   }
+                   else if( pixel > upper_bound )
+                   {
+                     return static_cast< OutputType >( std::numeric_limits< OutputType >
+                                                       ::max() );
+                   }
+                   else
+                   {
+                     return static_cast< OutputType >( ( pixel -
+                                                         lower_bound ) * scale );
+                   }
+                 } );
+}
+
+} // namespace <anonoymous>
+
+// ----------------------------------------------------------------------------
+/// Private implementation class
+class convert_image::priv
+{
+public:
+  priv()
+    : format{ "byte" }
+      , single_channel{ false }
+      , scale_factor{ 0.0 }
+      , random_greyscale{ 0.0 }
+      , percentile_norm{ -1.0 }
+  {
+  }
+
+  // Apply appropriate transforms
+  template < typename ipix_t > vil_image_view< ipix_t >
```

Comment:
The `template` spec should probably not be on the same line as the return type. (In both cases, the return plus name and argument list fit on the same line, but breaking after the return type is also okay; feel free to pick whichever you prefer.)

Reply:
(Ping?)

Reply:
I left this style because uncrustify regresses to it.

Reply:
I did update it in in the most recent commit. However, I'm still a bit wary of going against `uncrustify` for low-value changes. I'd hate to see someone stop running `uncrustify` because they know they're going to have to manually address a bunch of regressions like this.

Reply:
Sigh... that's my bad; on further consideration, I don't think `nl_template_func_decl` should be `remove`. Here:
```c++
  template < typename OutputType, typename InputType >
  void
  nonmax_suppression( vil_image_view< InputType > const& grad_i,
                      vil_image_view< InputType > const& grad_j,
                      vil_image_view< OutputType >& output_i,
                      vil_image_view< OutputType >& output_j );
```
...you must be ignoring uncrustify? (Because `nl_template_func_decl = remove` should be removing the newline there also, and seems to do so.)

Reply:
Nota bene: While it's true that uncrustify has bugs, sometimes the bugs are instead in the configuration :slightly_smiling_face:. (When you have a few hundred knobs to twiddle, it's bound to happen occasionally.)

Reply:
Yes, it looks like you're right about that. I agree the shown (non-uncrustify) style is better, so I guess I've been rather sporadic in which way I've gone. You'll notice the following function, `filter`, is styled as uncrustify suggests...so I'm just a mess.